Unit tests

all3fox opened the pull request:
I am having some problems with batavia so I decided to sidestep a little and just shuffle things around in `boolean.py`, will now be getting back to making batavia work.

A thing to watch out for in this PR: readme display on `pypi` website might  break because they use a rather strict and custom (afaik) parser for `.rst` and `.md` there. Once this lands, I will keep an eye on that and fix it if need be.

pombredanne commented:
Is the  __init__.py in https://github.com/bastikr/boolean.py/pull/70/files#diff-498cf53d35427897613cdfc4b76fc6ea really needed?

pombredanne commented on tests/test_not.py:
```diff
@@ -0,0 +1,154 @@
+        assert (~(a & a & a)).simplify() == (~(a & a & a)).simplify()
+        assert (~(a | a | a)).simplify() == (~(a | a | a)).simplify()
+
+    @pytest.mark.xfail(reason="a.cancel() should work but does not")
```
You have changed the logic of this test.
It was:
```
    def test_cancel(self):
        algebra = BooleanAlgebra()
        a = algebra.Symbol('a')
        self.assertEqual(~a, (~a).cancel())
        self.assertEqual(a, algebra.parse('~~a').cancel())
        self.assertEqual(~a, algebra.parse('~~~a').cancel())
        self.assertEqual(a, algebra.parse('~~~~a').cancel())
```

So I am not sure cancelling  a symbol makes any sense.

all3fox replied:
My thinking process there was:

> there is a separate test that deals exclusively with `parse` and also parsing is never easy, so let's throw it out of this test and just use python objects

I also started a separate `test_parse.py` later, moving a small portion of `test_parse` from `OtherTestCase` . The move is not finished yet though.

Resolution: I will write a second version of `test_cancel` that uses `parse` then.

all3fox commented on test-requirements.txt:
```diff
@@ -1 +1,4 @@
 tox==2.7.0
+pytest==3.1.2
```
No, not really. Just in the spirit of exactly pinning dependencies. Later on, one could outsource dependency updating to a bot like https://pyup.io/ which will create PRs when a dependency could be updated. However, can change to `3.1.x` or even  `3.x.x` (not sure about the exact syntax though)

all3fox commented on setup.py:
```diff
@@ -5,40 +5,23 @@
-'''
-
+with open('README.md') as readme:
+    long_description = readme.read()
```
That is true, will rewrite the readme in `.rst` and check it then.

all3fox commented on tests/test_not.py:
```diff
@@ -0,0 +1,154 @@
+
+        assert a == a.simplify()
+        assert a == (~~a).simplify()
+        assert a == (~~ ~~a).simplify()
```
Actually, no, that space was intentional. It was there before:

https://github.com/bastikr/boolean.py/pull/70/commits/bfb0b31dbb2ced65edf6f64d42c7171c2daa2475#diff-b03f7319ce9e9d763799954c9b76efa9L241

all3fox commented:
Sure, I can first do the changes in the same file and in the same order so that it would be easier to review. However, please note that you could:

- instead of clicking on "Files changed" to see all the changes at once, click on the individual commits. That pulls up the diff for the exact test I was moving around/modifying.
- do a similar thing locally in console by `git diff hash0..hash1`. If you pick hashes of consecutive commits, you see exactly the difference between them.

Anyway, I totally understand that this PR is big/hard to follow. As you suggest, I can use it to quick-start another PR that would first do the changes and then (once you've seen the exact changes), splits the files.

all3fox commented:
> Is the init.py in https://github.com/bastikr/boolean.py/pull/70/files#diff-498cf53d35427897613cdfc4b76fc6ea really needed?

Removing the file and runnign `pytest` gives an import error. Tell me if you do not like the empty `__init__.py`, maybe it is possible to remove it and modify the imports somehow.

all3fox commented:
And I merged back the `test_symbol.py`, `test_not.py`, and `test_base_element.py`.

I did not merge back `mock_advanced_algebra.py` and `mock_custom_algebra.py` because the whole point was to put those classes into separate `mock` files.

I also did *not* merge back `test_parse.py` because those are *new* tests for the `parse` functionality, so I am writing them in a separate file.

Also rewrote `README.md` as `README.rst`, added some badges and checked it with `restview` and `python setup.py check --restructuredtext` -- seems ok.

all3fox commented:
Everything has been addressed, I am removing the `[WIP]` prefix and waiting for a review.

pombredanne commented on setup.py:
```diff
@@ -5,40 +5,23 @@
     packages=find_packages(),
     include_package_data=True,
     zip_safe=False,
-    test_loader='unittest:TestLoader',
```
The previous test suite could run equally well with or without pytest. Any specil reason to make py.test a hard requirement for tests?

ferdnyc replied:
(I know this is a _very_ old PR, but...)

The use of `@pytest` decorators makes pytest a hard requirement, full stop. (And I agree with @alisianoi: it's better _not_ to support multiple test frameworks, which lead to everyone executing the tests in subtly different ways. Do that, you've stopped testing your code; now you're compatibility-testing the frameworks instead.)

`pytest-runner` [is deprecated](https://pypi.org/project/pytest-runner/#description). (As are any use of `test_runner`, `test_suite`, `tests_require`, etc. in `setup.py`. `setup.py` _as a whole_ is deprecated, of course, but its test args are "super double deprecated".) These days, `tox` would be a better choice as test runner.

pombredanne commented on tests/test_boolean.py:
```diff
@@ -324,217 +195,419 @@ def test_parse_side_by_side_symbols_with_parens_raise_exception(self):
+        algebra = BooleanAlgebra()
+
+        assert     algebra.Symbol('a') == algebra.Symbol('a')
+        assert not algebra.Symbol('a') == algebra.Symbol('b')
```
same comment as above. I am not 100% sure that `not ==` is the same as `!=` here because of the (and, or, not) operator overloading

pombredanne commented on tests/test_boolean.py:
```diff
@@ -324,217 +195,419 @@ def test_parse_side_by_side_symbols_with_parens_raise_exception(self):
+    def test_symbols_eq_ne(self):
+        algebra = BooleanAlgebra()
+
+        symbols0 = [
```
I tend to prefer many small tests methods that a loop... Now  if we keep a loop, this still might need exploding in different purpose tests?   I find it a tad hard to read what the tests are really doing

all3fox replied:
I see, this is actually something that I like to do: a few manual single unit tests above (like the `test_symbol_ne*` above) and then a final big test (that I also sometimes write as a test generator). If this big test starts to fail, then I find out which exact test case is failing and factor it out into yet another small, separate `test_symbol_ne` test. Simply factoring everything into nice small tests is too much typing and repetetive test names :)

all3fox replied:
But can do if you want, sure

pombredanne commented on tests/test_boolean.py:
```diff
@@ -324,217 +195,419 @@ def test_parse_side_by_side_symbols_with_parens_raise_exception(self):
-    def test_demorgan(self):
+        assert a == a.simplify()
+        assert a == (~~a).simplify()
+        assert a == (~~ ~~a).simplify()
```
I think the space  in the middle of the `~~ ~~` here and there is likely a typo that I introduced in the past from some editor reformatting.

pombredanne commented on tests/test_boolean.py:
```diff
@@ -324,217 +195,419 @@ def test_parse_side_by_side_symbols_with_parens_raise_exception(self):
+        assert parse('~(~a & b)').demorgan() == parse('a | ~b')
+        assert parse('~(a & ~b)').demorgan() == parse('~a | b')
+
+    # TODO: enforced order is obscure, must explain
```
You mean the precedence of operators? I agree, this is also obscure to me

all3fox replied:
I meant the lines `585` to `595` below that use the `<` and `>` comparison operators. Is that what you meant?

pombredanne commented on tests/test_parse.py:
```diff
@@ -0,0 +1,167 @@
+            assert algebra.parse('((' + x + '))') == algebra.TRUE
+            assert algebra.parse('(((' + x + ')))') == algebra.TRUE
+
+    @pytest.mark.xfail(reason='IndexError indicates parsing problem')
```
Can you elaborate on this?

all3fox replied:
If you remove the `xfail` decorator, the parser breaks and the whole thing crashes with an `IndexError`. That is not supposed to happen because the parser was designed to throw custom errors when there is a problem. So, you either throw your own errors or generic errors, not both (imho). Since it starts to throw generic errors all over sudden means there is likely some corner case that was not considered.

pombredanne commented on tests/test_parse.py:
```diff
@@ -0,0 +1,167 @@
+            with pytest.raises(ParseError):
+                parse(expression)
+
+    @pytest.mark.xfail(reason='IndexError indicates parsing problem')
```
Same as above. Which index error?

all3fox commented on setup.py:
```diff
@@ -5,40 +5,23 @@
     packages=find_packages(),
     include_package_data=True,
     zip_safe=False,
-    test_loader='unittest:TestLoader',
```
Just to force everybody to use the same thing. The more options you have, the more (compatibility?) things you have to keep in mind. With just one thing there is only one way: `pip install -r requirements.txt; pytest`

all3fox commented on tests/test_boolean.py:
```diff
@@ -324,217 +195,419 @@ def test_parse_side_by_side_symbols_with_parens_raise_exception(self):
+        assert algebra.FALSE == algebra.FALSE
+
+        assert     algebra.TRUE != algebra.FALSE
+        assert not algebra.TRUE == algebra.FALSE
```
Was writing like this first:
```
assert not (algebra.TRUE == algebra.FALSE)
```

and then noticed that the `()` can be dropped, so I dropped them. Will rewrite as:

```
        assert      algebra.TRUE != algebra.FALSE
        assert not (algebra.TRUE == algebra.FALSE)
```

all3fox commented on tests/test_boolean.py:
```diff
@@ -324,217 +195,419 @@ def test_parse_side_by_side_symbols_with_parens_raise_exception(self):
+        assert repr(algebra.TRUE) == 'TRUE'
+        assert repr(algebra.FALSE) == 'FALSE'
 
-    def test_init(self):
```
As far as I could tell, these were not tests: no assert statements of any kind. If they were to fail, that would be ugly tracebacks, not test results. Besides, that is just assignment under the hood, so don't see what is being tested.

On the other hand, the `init` part of the test name is confusing and might suggest that this is part of some test setup, which it is not. So, removed.

all3fox commented on tests/test_boolean.py:
```diff
@@ -324,217 +195,419 @@ def test_parse_side_by_side_symbols_with_parens_raise_exception(self):
+
+    def test_symbols_eq_0(self):
+        algebra = BooleanAlgebra()
+
```
Actually, not really. This shows three parts of the unit test: setup, run and assertion. But can make it three lines, no spaces if you want.

all3fox replied:
And I could not resist:

<a href="https://imgflip.com/i/1rs9jw"><img src="https://i.imgflip.com/1rs9jw.jpg" title="made at imgflip.com"/></a>

all3fox commented on tests/test_boolean.py:
```diff
@@ -324,217 +195,419 @@ def test_parse_side_by_side_symbols_with_parens_raise_exception(self):
+    def test_true_and_false(self):
         algebra = BooleanAlgebra()
-        self.assertEqual(algebra.TRUE, algebra.TRUE)
-        BaseElement()
```
Oh, you mean `BaseElement()`, the empty call. So, it should test that an empty call does not trigger any sort of exception?

This is the same type of test that I just removed from https://github.com/bastikr/boolean.py/pull/70/files/29e66f0e82244f90608eaad9739cf0271cb9a983#r125046762

I will look up how to write a unit test that "makes sure this is legal" (or tell me if you know how) and bring this and the linked tests from other comment back.

all3fox commented:
Ok, I have addressed:

1. The two places where I removed tests. They are now back with explicit exception checks.
2. The `not` and missing `()`, I have written explicit `()`
3. Squashed the spaces in `~ ~~` into `~~~`

For you to decide:

1. why pytest: https://github.com/bastikr/boolean.py/pull/70#discussion_r125045133
2. (not) to split the big one: https://github.com/bastikr/boolean.py/pull/70#discussion_r125078228
3. the xfail'ed `IndexError`: https://github.com/bastikr/boolean.py/pull/70#discussion_r125079649

Also, when answering your questions, I misclicked the "Start Review" button (or something) and my last three comments ended up as a separate review. I cannot "dismiss" it, so if you could that would be great. Also, I am new to the review mechanic on github so please tell me when I am doing things wrong.

Will be removing the `[WIP]` and waiting for your review.
